### DISCUSSION TITLE
[API Request] GetAlbumList2 endpoint: sorting and filtering should not be mutually exclusive

### DISCUSSION THREAD

**Discussion by dweymouth** [opensubsonic/open-subsonic-api#7] [API requests]
### Type of change

API tweak

### Proposal description

Currently, with the `GetAlbumList2` endpoint, you can retrieve a list of all albums, paginated and sorted (by recently added, name a-z, etc.), OR you can retrieve a filtered list of albums (e.g. byGenre, by year range), but you can't specify a sort order on the result. Tweaking/extending the API to allow a sort and filter specified together would enable scenarios like "show recently added Rock albums"

### Backward compatibility impact

If the API is called the way it is called today, i.e. specifying `type=byGenre` or `byYear` then the behavior is the same as the status quo

### Backward compatibility

- [X] No backward compatibility impact.

### API details

I propose tweaking the `GetAlbumList2` endpoint to accept the `fromYear`, `toYear`, and `genre` options without mandating that you also set `type=byYear or byGenre`. E.g. a request with `genre=Classical`, `fromYear=1990`, toYear=`2023`, `type=recent` should get all classical albums from 1990-2003, and return the result in recently added sort order

**Comment by gravelld** [comment]
Agee, but I think I'd expect this is how it should work in the first place, given http://www.subsonic.org/pages/api.jsp#getAlbumList2 . "_can_" is used, rather than "_must_" or "_shall_" which makes it almost sound optional, and no reference is made to `type` from the filtering parameters.

**Reply by dweymouth** [reply]
@opensubsonic/servers For each server, what is the implementation status of supporting `fromYear` `toYear` and `genre` alongside `type=[recent|frequent|etc]`? (ie allowing sorting the results of a filtered album list). If all OpenSubsonic servers already support this we can probably just add the bolded to the documentation of the `type` parameter for `getAlbumList2`:

> The list type. Must be one of the following: random, newest, highest, frequent, recent. Since [1.8.0](https://opensubsonic.netlify.app/docs/subsonic-versions) you can also use alphabeticalByName or alphabeticalByArtist to page through all albums alphabetically, and starred to retrieve starred albums. Since [1.10.1](https://opensubsonic.netlify.app/docs/subsonic-versions) you can use byYear and byGenre to list albums in a given year range or genre. **For all OpenSubsonic servers, fromYear, toYear, and genre can be specified without requiring type=byYear or byGenre.**

**Reply by epoupon** [reply]
Maybe we could even make the type param multi valued? This way we can process multiple filters and sort. Too bad some combinations make no sense though.

**Reply by sentriz** [reply]
this would suit me. in the case of gonic it takes the filter/sort string and maps it to SQL clauses so it would be fine to handlde multiple of them

**Reply by dweymouth** [reply]
That is a separate discussion on how to extend the API. I was trying to see if we could get a quick win by just adding the bolded language to the documentation before we actually go about defining a proper OpenSubsonic extension for album list. As @gravelld said the original API was a bit underspecified and a lot of servers may have already implemented it that way rather than ignoring the `genre=foo` parameter if `type` was not also `byGenre`

If every single OpenSubsonic server has already supported using the existing parameters this way from the get-go - to sort and filter together - we don't even need to bother with an OpenSubsonicExtension definition since clients would know any server that returns `openSubsonic=true` in its subsonic-response would support it.
